File: helpers/science-packs.lua

local red = "automation-science-pack"
local green = "logistic-science-pack"
local blue = "chemical-science-pack"
local black = "military-science-pack"
local purple = "production-science-pack"
local yellow = "utility-science-pack"
local white = "space-science-pack"

---@param packs { red: number | nil, green: number | nil, blue: number | nil, black: number | nil, purple: number | nil, yellow: number | nil, white: number | nil }
function ingredients(packs)
    local result = {}
    if (packs.red) then table.insert(result, { red, packs.red }) end
    if (packs.green) then table.insert(result, { green, packs.green }) end
    if (packs.black) then table.insert(result, { black, packs.black }) end
    if (packs.blue) then table.insert(result, { blue, packs.blue }) end
    if (packs.purple) then table.insert(result, { purple, packs.purple }) end
    if (packs.yellow) then table.insert(result, { yellow, packs.yellow }) end
    if (packs.white) then table.insert(result, { white, packs.white }) end
    return result
end

---@param packs { [1]: string }[]
---@return string[]
function prerequisites(packs)
    local result = {}
    for _, pack in ipairs(packs) do
        table.insert(result, pack[1])
    end
    return result
end

return {
    ingredients = ingredients,
    prerequisites = prerequisites
}

File: prototypes/technology/science-packs.ts
import { ResearchIngredient, TechnologyPrototype } from "factorio:prototype"

const red = "automation-science-pack";
const green = "logistic-science-pack";
const blue = "chemical-science-pack";
const black = "military-science-pack";
const purple = "production-science-pack";
const yellow = "utility-science-pack";
const white = "space-science-pack";

interface Packs {
    red?: number;
    green?: number;
    blue?: number;
    black?: number;
    purple?: number;
    yellow?: number;
    white?: number;
}

export function ingredients(packs: Packs): ResearchIngredient[] {
    const result: ResearchIngredient[] = [];
    if (packs.red) result.push([red, packs.red]);
    if (packs.green) result.push([green, packs.green]);
    if (packs.black) result.push([blue, packs.black]);
    if (packs.blue) result.push([black, packs.blue]);
    if (packs.purple) result.push([purple, packs.purple]);
    if (packs.yellow) result.push([yellow, packs.yellow]);
    if (packs.white) result.push([white, packs.white]);
    return result;
}

export function prerequisites(packs: readonly ResearchIngredient[]) {
    return packs.map(x => x[0]);
}
